e2e/terraform/compute.tf

```diff
@@ -0,0 +1,124 @@
+  count                  = "${var.server_count}"
+
+  # Instance tags
+  tags {
+    Name           = "${local.random_name}-server-${count.index}"
+    ConsulAutoJoin = "auto-join"
+  }
+
+  user_data            = "${data.template_file.user_data_server.rendered}"
+  iam_instance_profile = "${aws_iam_instance_profile.instance_profile.name}"
+
+  provisioner "file" {
+    content     = "${file("${path.root}/configs/${var.indexed == false ? "server.hcl" : "indexed/server-${count.index}.hcl"}")}"
+    destination = "/tmp/server.hcl"
+
+    connection {
+      user        = "ubuntu"
+      private_key = "${module.keys.private_key_pem}"
+    }
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "aws s3 cp s3://nomad-team-test-binary/builds-oss/${var.nomad_sha}.tar.gz nomad.tar.gz",
+      "sudo cp /ops/shared/config/nomad.service /etc/systemd/system/nomad.service",
+      "sudo tar -zxvf nomad.tar.gz -C /usr/local/bin/",
+      "sudo cp /tmp/server.hcl /etc/nomad.d/nomad.hcl",
+      "sudo chmod 0755 /usr/local/bin/nomad",
+      "sudo chown root:root /usr/local/bin/nomad",
+      "sudo systemctl start nomad.service"
+    ]
+
+    connection {
+      user        = "ubuntu"
+      private_key = "${module.keys.private_key_pem}"
+    }
+  }
+}
+
+resource "aws_instance" "client" {
+  ami                    = "${data.aws_ami.main.image_id}"
+  instance_type          = "${var.instance_type}"
+  key_name               = "${module.keys.key_name}"
+  vpc_security_group_ids = ["${aws_security_group.primary.id}"]
+  count                  = "${var.client_count}"
+  depends_on             = ["aws_instance.server"]
+
+  # Instance tags
+  tags {
+    Name           = "${local.random_name}-client-${count.index}"
+    ConsulAutoJoin = "auto-join"
+  }
+
+  ebs_block_device =  {
+    device_name                 = "/dev/xvdd"
+    volume_type                 = "gp2"
+    volume_size                 = "50"
+    delete_on_termination       = "true"
+  }
+
+  user_data            = "${element(data.template_file.user_data_client.*.rendered, count.index)}"
+  iam_instance_profile = "${aws_iam_instance_profile.instance_profile.name}"
+
+  provisioner "file" {
+    content     = "${file("${path.root}/configs/${var.indexed == false ? "client.hcl" : "indexed/client-${count.index}.hcl"}")}"
+    destination = "/tmp/client.hcl"
+
+    connection {
+      user        = "ubuntu"
+      private_key = "${module.keys.private_key_pem}"
+    }
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "aws s3 cp s3://nomad-team-test-binary/builds-oss/${var.nomad_sha}.tar.gz nomad.tar.gz",
+      "sudo tar -zxvf nomad.tar.gz -C /usr/local/bin/",
+      "sudo cp /ops/shared/config/nomad.service /etc/systemd/system/nomad.service",
+      "sudo cp /tmp/client.hcl /etc/nomad.d/nomad.hcl",
+      "sudo chmod 0755 /usr/local/bin/nomad",
+      "sudo chown root:root /usr/local/bin/nomad",
+      "sudo systemctl start nomad.service"
+    ]
```

Comment:
We should add an additional readiness check for Nomad here so Terraform provisioning doesn't return complete until Nomad is ready. I think you could be racing currently given the cloud-init scripts have a bunch of sleeps in them.

Not sure if you could do a bash one-liner or if there is a Nomad CLI command for readiness...

Reply:
yeah we need a wait for leader util function. I thought it made more sense as a util function to use inside our test framework rather than put it here as a precondition for terraform to return.

Reply:
That sounds like a good approach!